File: Assets/src/model/enemies/Wallflower.cs
using System.Linq;
using UnityEngine;

[System.Serializable]
[ObjectInfo(description: "Chases you.\nMust stick next to a wall.")]
public class Wallflower : AIActor {
  // public static Item HomeItem => new ItemPlaceableTile(typeof(Wall));
  public static Item HomeItem => new ItemWallflowerTendril();
  public Wallflower(Vector2Int pos) : base(pos) {
    faction = Faction.Enemy;
    hp = baseMaxHp = 2;
  }

  public static bool CanOccupy(Tile t) => t.CanBeOccupied() && t.floor.GetAdjacentTiles(t.pos).Any(n => n is Wall);

  protected override ActorTask GetNextTask() {
    var tethers = floor.GetCardinalNeighbors(pos).OfType<Wall>().ToList();

    if (!tethers.Any()) {
      // oh god! walk randomly until you are touching
      return new MoveRandomlyTask(this);
    }

    var player = GameModel.main.player;
    if (CanTargetPlayer()) {
      if (IsNextTo(player)) {
        return new AttackTask(this, player);
      }
    }

    var nextTethers = tethers
      .SelectMany(touchingWall => floor.GetCardinalNeighbors(touchingWall.pos, true).OfType<Wall>()).Distinct().ToList();
    var nextTetherOccupiableTiles = nextTethers
      .SelectMany(w => floor.GetAdjacentTiles(w.pos))
      .Distinct()
      .Where(t => CanOccupy(t) || t == this.tile).ToList();
    
    var adjacent = floor.GetAdjacentTiles(pos).ToList();
    var candidateTiles = nextTetherOccupiableTiles.Intersect(adjacent).ToList();

    // var candidateTiles = floor.GetAdjacentTiles(pos).Where(CanOccupy).ToList();

    var nextTile = candidateTiles.OrderBy((t) => t.DistanceTo(GameModel.main.player)).FirstOrDefault();

    if (nextTile == this.tile) {
      return new WaitTask(this, 1);
    } else if (nextTile != null) {
      return new MoveToTargetTask(this, nextTile.pos);
    } else {
      return new WaitTask(this, 1);
    }
  }

  internal override (int, int) BaseAttackDamage() => (1, 1);
}

[System.Serializable]
[ObjectInfo("wallflower-tendril", description: "Also makes a free attack on the creature directionally behind the one you attack.")]
public class ItemWallflowerTendril : EquippableItem, IWeapon, IActionPerformedHandler, IDurable {
  public (int, int) AttackSpread => (1, 1);
  public override EquipmentSlot slot => EquipmentSlot.Weapon;

    public int durability { get; set; }

    public int maxDurability => 24;

    public ItemWallflowerTendril() : base() {
      durability = maxDurability;
    }

    public void HandleActionPerformed(BaseAction final, BaseAction initial) {
    if (final is AttackBaseAction attack) {
      var target = attack.target;
      var attacker = attack.actor;
      var offset = target.pos - attacker.pos;
      if (target.IsNextTo(attacker)) {
        var posBehind = target.pos + offset;
        if (attacker.floor.InBounds(posBehind) && attacker.floor.bodies[posBehind] != null) {
          attacker.Perform(new AttackBaseAction(attacker, attacker.floor.bodies[posBehind], true));
        }
      }
    }
  }
}

File: web/src/model/enemies/Wallflower.ts
import { AIActor } from './AIActor';
import { ActorTask } from '../ActorTask';
import { AttackTask } from '../tasks/AttackTask';
import { MoveToTargetTask } from '../tasks/MoveToTargetTask';
import { MoveRandomlyTask } from '../tasks/MoveRandomlyTask';
import { WaitTask } from '../tasks/WaitTask';
import { Vector2Int } from '../../core/Vector2Int';
import { GameModelRef } from '../GameModelRef';
import { Wall } from '../Tile';
import { entityRegistry } from '../../generator/entityRegistry';
import type { Tile } from '../Tile';

/**
 * Chases you but must stay adjacent to walls.
 * Port of C# Wallflower.cs.
 */
export class Wallflower extends AIActor {
  constructor(pos: Vector2Int) {
    super(pos);
    this._hp = this._baseMaxHp = 2;
  }

  baseAttackDamage(): [number, number] {
    return [1, 1];
  }

  /** A tile can be occupied by a wallflower only if it's occupiable AND adjacent to a wall. */
  static canOccupy(t: Tile): boolean {
    if (!t.canBeOccupied()) return false;
    const adjacent = t.floor!.getAdjacentTiles(t.pos);
    return adjacent.some(n => n instanceof Wall);
  }

  protected getNextTask(): ActorTask {
    const floor = this.floor!;
    const tethers = floor.getCardinalNeighbors(this.pos).filter(t => t instanceof Wall);

    if (tethers.length === 0) {
      // Not touching a wall — walk randomly until we are
      return new MoveRandomlyTask(this);
    }

    const player = GameModelRef.main.player;
    if (this.canTargetPlayer()) {
      if (this.isNextTo(player)) {
        return new AttackTask(this, player);
      }
    }

    // Find walls reachable via wall-adjacent-to-wall chain (one step out)
    const nextTethers = new Set<Tile>();
    for (const touchingWall of tethers) {
      for (const neighbor of floor.getCardinalNeighbors(touchingWall.pos, true)) {
        if (neighbor instanceof Wall) {
          nextTethers.add(neighbor);
        }
      }
    }

    // Candidate tiles: adjacent to me, wallflower-occupiable, and cardinally next to a nextTether
    const candidateTiles = floor
      .getAdjacentTiles(this.pos)
      .filter(t => Wallflower.canOccupy(t))
      .filter(adjacent =>
        floor.getCardinalNeighbors(adjacent.pos).some(n => nextTethers.has(n)),
      );

    // Pick tile closest to player
    const playerPos = GameModelRef.main.player.pos;
    const nextTile = candidateTiles.length > 0
      ? candidateTiles.reduce((best, t) =>
          Vector2Int.distance(t.pos, playerPos) < Vector2Int.distance(best.pos, playerPos) ? t : best,
        )
      : null;

    if (nextTile != null && nextTile !== this.tile) {
      return new MoveToTargetTask(this, nextTile.pos);
    }
    return new WaitTask(this, 1);
  }
}

entityRegistry.register('Wallflower', Wallflower);
